### PR TITLE
Expose neighbor lists via library interface - Second iteration

### PR DESCRIPTION
**Pull request by rbberger**
**Summary**

Exposes neighbor lists via the C library interface and in the Python library.

Example (Python):
```python
L = lammps()
L.file('in.melt')

nlocal = L.extract_global("nlocal", 0)
nghost = L.extract_global("nghost", 0)
x = L.numpy.extract_atom_darray("x", nlocal+nghost, dim=3)
v = L.numpy.extract_atom_darray("v", nlocal+nghost, dim=3)
f = L.numpy.extract_atom_darray("f", nlocal+nghost, dim=3)

# return neighbor list object, which encapsulates access via library interface

mylist = L.find_pair_neighlist("lj/cut")

# return length of list
len(mylist)
# or
mylist.size

# get neighbors of list element 0
iatom, numneigh, neighs = mylist[0]
# ^        ^       ^
# |        |       | 
# |        |       + ------- neighbors of iatom (numpy arrray of integer indices)
# |        +-------------- number of neighbors of iatom
# +--------------------- index of atom that is element 0 in neighbor list

# equivalent:
iatom, numneigh, neighs = mylist.get(0)

# neighbor list can also be used as iterator
for iatom, numneigh, neighs in mylist:
      print(x[iatom], v[iatom], f[iatom], " : ",  numneigh, "Neighbors")
      for jatom in neighs:
        if jatom < nlocal:
            print("    * [LOCAL]", jatom, x[jatom], v[jatom], f[jatom])
        else:
            print("    * [GHOST]", jatom, x[jatom], v[jatom], f[jatom])

```

**Related Issues**

#1287

**Author(s)**

@rbberger (Temple U)

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**
yes

**Implementation Notes**

Adds the following library functions to the library interface:

```c++
// search for neighbor list for pair style, returns index
int lammps_find_pair_neighlist(void* ptr, char * style, int nsub, int request);

// search for neighbor list for fix, returns index
int lammps_find_fix_neighlist(void* ptr, char * id, int request);

// search for neighbor list for compute, returns index
int lammps_find_compute_neighlist(void* ptr, char * id, int request);

// get total size of neighbor list of given index
int lammps_neighlist_num_elements(void* ptr, int idx);

// return atom id, numneigh and neighbors of neighbor list element
void lammps_neighlist_element_neighbors(void * ptr, int idx, int element, int * iatom, int * numneigh, int ** neighbors);
```

The Python library interface uses these functions via CTypes. For ease of use they have been wrapped in a utility class called `NeighList`, which acts similar to a Python list. While the low level functions can be used, the `lammps` object provides the following utility functions which return a `NeighList` object. No arrays are copied, it uses numpy arrays that directly access the C memory blocks.

```Python
  def get_neighlist(self, idx):
        ....

  def find_pair_neighlist(self, style, nsub=0, request=0):
        ....

  def find_fix_neighlist(self, fixid, request=0):
        ...

  def find_compute_neighlist(self, computeid, request):
        ...
```
To enable the `find_XYZ_neighlist` functions on both the C and Python side, changes were necessary in the Neighbor List C++ code. This PR changes `NeighList` to include `requestor`, which is a pointer to the instance which asked for a neighbor list, and `requestor_type`, which is an enum to specify if it was a Pair, Fix or Compute. This was necessary because this information was only part of `NeighRequest`, which is thrown away after the list is created. I don't see a way of determining which neighbor list belongs to who without this change.

**Post Submission Checklist**

_Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply_

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

_Put any additional information here, attach relevant text or image files, and URLs to external sites (e.g. DOIs or webpages)_

### CODE REVIEW

File: src/library.cpp [hunk 1]
```diff
@@ -1727,3 +1730,133 @@ int lammps_get_last_error_message(void *ptr, char * buffer, int buffer_size) {
...
+  if(idx < 0 || idx >= neighbor->nlist) {
+    return -1;
+  }
+
+  NeighList * list = neighbor->lists[idx];
+  return list->inum;
+}
+
+/* ----------------------------------------------------------------------
+   Return atom index, number of neighbors and neighbor array for neighbor
```
Comment (sjplimp) [inline src/library.cpp]: This needs more careful doc - which are the inputs, which are the outputs?
What is element?
Is iatom a local index or atom ID?
Say that numneigh and neighbors are length returned by lammps_neighlist_num_elements.
Not clear why there are 2 separate methods for length and these returns.
Why not have this method just return the length as the function return?
Document that the length of these are not necessarily the Nlocal for that proc.

**Comment by akohlmey** [comment]
assigning to @sjplimp for further review and testing. please assign it back to me when satisfied.

**Comment by akohlmey** [comment]
@rbberger this is just a friendly reminder, that this PR is waiting for updates from you as requested by @sjplimp

File: src/library.cpp (hunk 1 above)
Comment (rbberger) [reply src/library.cpp]: @sjplimp `lammps_neighlist_num_elements` returns `inum`, which is the number of elements in the neighbor list. That is not the number of neighbors of a single atom (aka `numneigh`).

**Comment by rbberger** [comment]
@sjplimp please have another look. Take a closer look at the generated `Python_library.html` and let me know what you think.

**Comment by sjplimp** [comment]
@rbberger I like the added examples/python script.  I like the (added?) Python library doc page.
However I'm not seeing any new entries for the neighbor methods on that page or the Howto_library.rst page?  Do you want me to add them?  I can also add a few more comments in library.cpp where I think users may look first.

**Comment by sjplimp** [comment]
@rbberger Ah, now I did a make html on the doc pages and see the Neigh list content added at the end of Python_library.html.  I didn't see how that was done in the *.rst file.  So that content is coming from library.cpp itself?  Is the plan to do that for all the methods, both in Python_library and in Howto_library?  Are you wanting to wait on this PR until that is done?

**Comment by rbberger** [comment]
@sjplimp so what I did is getting source documentation from Python code (`lammps.py`) into the RST files. This is done through what is called `autodoc`. To achieve something similar with comments in C++ code, I would first have to generate an XML file using Doxygen. That would add another build dependency to the documentation build.

If you approve, I could then add markup in `Howto_library.rst` using something called [Breathe](https://breathe.readthedocs.io/en/latest/). This will look similar to what you now see with Python code.

The alternative is to write it manually in `Howto_library.rst`, which is more prone to being outdated.

**Comment by rbberger** [comment]
@sjplimp the Doxygen + Breathe + Sphinx integration can now be seen in `Howto_library.html`. Please let me know what you think. I've limited it to `library.h` and `library.cpp`.

**Comment by sjplimp** [comment]
@rbberger What I don't understand about this syntax for either Python or cpp files is how much flexibility there is to adding comments or explanations.  In the current Howto_library.rst I think the section that explains the lib methods in paragraphs is useful in addition to the prototypes.  Similiarly there are text explanations in library.cpp beyond just one line per argument.  Maybe  if you could send me the 2 actual HTML pages and corresponding Python and library.cpp files, I would understand better.

**Comment by rbberger** [comment]
Cherry-picking  5d7ce839a162fd16448689685840b2a3738024fe and 3353bffb7211d5521f365ffacd491707b4a958a5 into PR #1787  so they don't get held up longer.

**Comment by sjplimp** [comment]
@rberger I'm not clear if you are waiting on me for something on this one.  I am fine with the code additions.  Just a good way to document both the lib and Python interface is what I am still unclear on.
In my view it doesn't have to be fully automated.  I'm more interested in clarity for users, so not
all the doc info necessarily needs to be in the code.

**Comment by rbberger** [comment]
@sjplimp If you don't want Doxygen and Breathe to document the library interface, say so and I'll remove the commit containing the changes.

**Comment by rbberger** [comment]
removing the Breathe/Doxygen changes for now. They will become their own PR.

**Comment by rbberger** [comment]
@sjplimp please approve the PR so I can merge.

**Pull request merged**